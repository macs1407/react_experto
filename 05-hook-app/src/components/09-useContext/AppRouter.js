import React from 'react';
import {BrowserRouter as Router, Route, Switch, Redirect} from 'react-router-dom';
import {AboutScreen} from './AboutScreen';
import {LoginScreen} from './LoginScreen';
import {HomeScreen} from './HomeScreen';
import { NavBar } from './NavBar';

export const AppRouter = () => {
    // Encargado de manejar las rutas
    return (
        <Router>
            <div>
                <NavBar />
                {/* Dependiendo de la condicion manda a una pagina */}
                <div className="container">
                    <Switch>
                        <Route exact path="/" component={HomeScreen} />
                        <Route exact path="/about" component={AboutScreen} />
                        <Route exact path="/login" component={LoginScreen} /> 
                        {/* Si no se encuentra la ruta en el switch se manda a home */}
                        <Redirect to="/" />                    
                    </Switch>
                </div>                
            </div>
        </Router>
    )
}
