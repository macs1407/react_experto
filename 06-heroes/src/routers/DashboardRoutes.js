import React from 'react';
import { Switch, Route, Redirect } from 'react-router-dom';
import { MarvelScreen } from '../components/marvel/MarvelScreen';
import { HeroesScreen } from '../components/heroes/HeroesScreen';
import { DcScreen } from '../components/dc/DcScreen';
import { Navbar } from '../components/ui/NavBar';
import { SearchScreen } from '../components/searchs/SearchScreen';


export const DashboardRoutes = () => {
    return (
        <>
            <Navbar />
            <div className="container mt-2">
                <Switch>
                    <Route exact path="/marvel" component={MarvelScreen}/>
                    <Route exact path="/heroe/:heroeId" component={HeroesScreen}/>
                    <Route exact path="/dc" component={DcScreen}/>
                    <Route exact path="/buscar" component={SearchScreen}/>
                    {/* Si no encuentra ninguno de los anteriores llama a /marvel */}
                    <Redirect to="/marvel" />
                </Switch>
            </div>
        </>
    )
}
