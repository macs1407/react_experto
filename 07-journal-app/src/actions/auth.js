import { types } from "../types/types";
import {firebase, googleAuthProvider} from '../firebase/firebase-config';
import { startLoading,finishLoading } from "./ui";
import Swal from 'sweetalert2';
import { noteLogout } from "./notes";

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});
// EL DISPTACH SIEMPRE SE ENCARGA DE ENVIAR LAS ACCIONES A LOS REDUCERS
// Crear peticion asincronica
export const startLoginEmailPassword = (email, password)=>{
    return (dispatch)=>{
        dispatch(startLoading());
        firebase.auth().signInWithEmailAndPassword(email, password)
        .then(({user})=>{
            
            setTimeout(()=>{
                dispatch(login(user.uid, user.displayName));
            },5000);            
            dispatch(finishLoading());
        }).catch(error=>{
            console.log(error);
            dispatch(finishLoading());
            Swal.fire('Error', error.message, 'error');
        });
    }
}
// Login de google 018000422222
export const startGoogleLogin = ()=>{
    return (dispatch)=>{
        firebase.auth().signInWithPopup(googleAuthProvider)
                .then(({user})=>{
                    dispatch(login(user.uid, user.displayName));
                });
    }
} 

export const startRegisterWithEmailPassword =(email, password, name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email, password)
        .then(async ({user})=>{
            await user.updateProfile({displayName:name});
            dispatch(login(user.uid, user.displayName));
        }).catch(error=>{
            console.log(error);
            Swal.fire('Error', error.message, 'error');
        })
    }
}

export const startLogout = ()=>{
    return async(dispatch)=>{
        await firebase.auth().signOut();
        dispatch(logout());
        // El reducer sabe que hacer aqui por que todo esta centrado en un state
        dispatch(noteLogout());
    }
}

export const logout = ()=>({
    type:types.logout
});