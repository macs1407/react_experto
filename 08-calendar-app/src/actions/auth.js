import { types } from "../types/types";
import { fetchSinToken, fetchConToken } from "../helpers/fetch";
import Swal from 'sweetalert2';
import { eventClearAll } from "./events";

export const startLogin = (email, password)=>{
    // Disparar el dispatch gracias a thunk
    return async(dispatch)=>{
        const resp = await fetchSinToken('auth', {email, password}, 'POST');
        const body = await resp.json();
        // Si todo sale bien
        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            // Hacer el dispatch para el store
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

const login = (user)=>({
    type:types.authLogin,
    payload:user
});

export const startRegister = (email, name, password)=>{
    return async(dispatch)=>{
        const resp = await fetchSinToken('auth/new', {email, name, password}, 'POST');
        const body = await resp.json();
        // Si todo sale bien
        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            // Hacer el dispatch
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }else {
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

export const startChecking = ()=>{
    return async(dispatch)=>{
        // Enviar la peticion
        const resp = await fetchConToken('auth/renew');
        // Obtener el json
        const body = await resp.json();
        console.log(body);
        // Si todo sale bien
        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            // Hacer el dispatch
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }else {
            // Disparar accion si el token es invalido
            dispatch(checkingFinish());
        }
    }
}

const checkingFinish = ()=>({
    type:types.authCheckingFinish
});

const logout = ()=>({
    type:types.authLogout
});

export const startLogout = ()=>{
    return(dispatch)=>{
        // remover el token
        localStorage.removeItem('token');
        dispatch(eventClearAll());
        dispatch(logout());        
    }
}