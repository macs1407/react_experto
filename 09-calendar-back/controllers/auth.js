// Obntener el tipado
const {response} = require('express');
const Usuario = require('../models/Usuario');
const bcrypt = require('bcryptjs');
const {generarJwt} = require('../helpers/jwt');

// Logica que lleva todo el control de las rutas
// resp = express.response se hace esto para que no se pierda el autocompletado
const crearUsuario = async(req, resp = response)=>{
    // req = solicitud, resp = respuesta que se devuelve
    const {email,password} = req.body;
    // Validar que el email no exista

    try{
        // Todos las consultas de mongo son promesas
        let usuario = await Usuario.findOne({email:email});
        // Si no existe ese correo electropnico
        if(usuario){
            return resp.status(400).json({
                ok:true,
                msg:'El usuario ya existe con ese correo'
            });
        } else{
            usuario = new Usuario(req.body);
            // Encriptar contraseña
            const salt = bcrypt.genSaltSync(10);
            usuario.password = bcrypt.hashSync(password, salt);
            // Grabar en usuario de base de datos
            await usuario.save();
            // Generar token
            const token = await generarJwt(usuario.id, usuario.name);
            return resp.status(201).json({
                ok:true,
                uid:usuario.id,
                nane:usuario.name,
                token
            });
        }
        
    }catch(error){
        return resp.status(500).json({
            ok:false,
            msg:'Ocurrio un error, hable con el administrador'
        })
    }
}

const loginUsuario = async(req, resp = response)=>{
    const {email, password} = req.body;
    // Manejo de errores con express validator
    try {
        let usuario = await Usuario.findOne({email:email});
        if(!usuario){
            return resp.status(400).json({
                ok:false,
                msg:"Email no existe"
            });
        } else {
            // Confirmar las contraseñas
            const validPassword = bcrypt.compareSync(password, usuario.password);
            if(!validPassword){
                return resp.status(400).json({
                    ok:false,
                    msg:"Contraseña incorrecta"
                })
            }
            // Generar token
            const token = await generarJwt(usuario.id, usuario.name);
            console.log(usuario);
            return resp.status(200).json({
                ok:true,
                uid:usuario.id,
                name:usuario.name,
                token
            })
        }
    }catch(error){
        return resp.status(500).json({
            ok:false,
            msg:'Ocurrio un error, hable con el administrador '+error
        })
    }
}

const revalidarToken = async(req, resp = response)=>{
    // En la peticion ya biene el uid y el name por que pasa por el validar-token y este setea esos valores
    const {uid, name} = req;
    // generar un nuevo token
    const token = await generarJwt(uid, name);

    resp.json({
        ok:true,
        uid, 
        name,
        token
    })
}


module.exports = {
    crearUsuario,
    loginUsuario,
    revalidarToken
}