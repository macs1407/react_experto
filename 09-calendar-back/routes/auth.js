// Rutas de usuario host + /api/auth
const express = require('express');
const router = express.Router();
const {check} = require('express-validator');
const {validarJwt} = require('../middlewares/validar-token');

const {crearUsuario, loginUsuario, revalidarToken} = require('../controllers/auth');
const {validarCampos} = require('../middlewares/validar-campos');

router.post('/new'
            // Middleware de express-validator
            ,[
                check('name', 'el nombre es obligatorio').not().isEmpty(),
                check('email', 'el email es obligatorio').not().isEmpty(),
                check('email', 'el email es invalido').isEmail(),
                check('password', 'la conatraseña es obligatoria').not().isEmpty(),
                check('password', 'la conatraseña debe contener almenos 6 caracteres').isLength({min:6}),
                validarCampos
            ]
            ,crearUsuario);

router.post('/'
            // Middleware de express-validator
            ,[
                check('email', 'el email es obligatorio').not().isEmpty(),
                check('email', 'el email es invalido').isEmail(),
                check('password', 'la conatraseña es obligatoria').not().isEmpty(),
                validarCampos
            ]            
            ,loginUsuario);

router.get('/renew', validarJwt, revalidarToken);

module.exports = router;
