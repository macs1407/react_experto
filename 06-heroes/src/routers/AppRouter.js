import React, { useContext } from 'react';
import {
    BrowserRouter as Router,
    Switch,
    Route
  } from "react-router-dom";
import { LoginScreen } from '../components/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';
import { PrivateRoute } from './PrivateRoute';
import { AuthContext } from '../auth/AuthContext';
import { PublicRoute } from './PublicRoute';

export const AppRouter = () => {
    const {user} = useContext(AuthContext);
    console.log(user);
    return (
        <Router>
        <div>
          <Switch>
            <PublicRoute 
              isAuthenticate={user.logged} 
              exact 
              path="/login" 
              component={LoginScreen} />
            <PrivateRoute 
              isAuthenticate={user.logged} 
              path="/" 
              component={DashboardRoutes} 
            />
          </Switch>
        </div>
      </Router>
    )
}
