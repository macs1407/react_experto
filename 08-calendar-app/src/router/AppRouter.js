import React, { useEffect } from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect
} from "react-router-dom";
import { LoginScreen } from "../components/auth/LoginScreen";
import { CalendarScreen } from "../components/calendar/CalendarScreen";
import { useDispatch, useSelector } from "react-redux";
import { startChecking } from "../actions/auth";
import { PrivateRoute } from "./PrivateRoute";
import { PublicRoute } from "./PublicRoute";

export const AppRouter = () => {
    const dispatch = useDispatch();

    const {checking, uid} = useSelector(state => state.auth);

    // Esto nos sirve para saber si esta logueado o no y si tien un token valido
    // haciendo esto automaticamente logueara en la aplicacion
    useEffect(() => {
        dispatch(startChecking());
    }, [dispatch]);
    
    // Si esta cargando
    if(checking){
        return <h5>cargando......</h5>
    }

    // No se debe mostrar nada hasta que el ckecking este en false
    return (
        <Router>
            <div>
                <Switch>
                    <PublicRoute isAuthenticate={!!uid} exact path="/login" component={LoginScreen} />
                    <PrivateRoute isAuthenticate={!!uid} exact path="/" component={CalendarScreen} />
                    <Redirect to="/" />
                </Switch>
            </div>
        </Router>
    )
}
