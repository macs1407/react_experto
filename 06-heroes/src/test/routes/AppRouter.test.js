import React from 'react';
import '@testing-library/jest-dom';
import {mount} from 'enzyme';
import { AppRouter } from '../../routers/AppRouter';
import { AuthContext } from '../../auth/AuthContext';
import { MemoryRouter } from 'react-router-dom';

describe('Pruebas en AppRouter', () => {
   const contexto = {
       dispatch: jest.fn(),
       user:{
           logged:false
       }
   }

    test('Debe de mostrar el componente si esta autenticado y guardar en localStorage ', () => {
        const wraper = mount (<AuthContext.Provider value={contexto}>
                                    <AppRouter />
                                </AuthContext.Provider>);
        expect(wraper).toMatchSnapshot();
    });    

    test('Debe de mostrar el componente de marvel si esta autenticado ', () => {
        const contexto = {
            dispatch: jest.fn(),
            user:{
                name:"maikol",
                logged:true
            }
        }
        const wraper = mount (<AuthContext.Provider value={contexto}>
            <AppRouter />
        </AuthContext.Provider>);
        console.log(wraper.html());
        expect(wraper.find('.navbar').exists()).toBe(true);
    })
    
})

