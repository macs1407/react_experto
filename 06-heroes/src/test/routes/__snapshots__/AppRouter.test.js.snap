// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AppRouter Debe de mostrar el componente si esta autenticado y guardar en localStorage  1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      Login Screen
    </h1>
    <hr />
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Login
    </button>
  </div>
</div>
`;
